### PR TITLE
Bug 470070 - Importing multi-module project with linked src folders does not create links

### PR DESCRIPTION
**Pull request by donat**
Create linked sources if the model contains it
Create linked resource only if it doesn't exist
Use linked resources as source folders if specified
Filter linked resources to work with folders only
Refactor source code updater in order to have better exception handling

### CODE REVIEW

**Comment by eclipsewebmaster** [comment]
Issue tracker reference:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=470070

**Comment by donat** [comment]
This is an extension of PR #126 which should be closed when this one gets merged.

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/SourceFolderUpdater.java [hunk 1]
```diff
@@ -76,43 +80,70 @@ private void updateClasspath(IProgressMonitor monitor) throws JavaModelException
...
-        // collect the paths of all source folders of the new Gradle model and keep any user-defined filters
-        List<IClasspathEntry> sourceFolders = FluentIterable.from(this.sourceFolders).transform(new Function<OmniEclipseSourceDirectory, IClasspathEntry>() {
+        // collect the paths of all source folders of the new Gradle model and keep any user-defined
+        // filters
+        ImmutableList.Builder<IClasspathEntry> sourceFolderEntries = ImmutableList.<IClasspathEntry>builder();
+        for (OmniEclipseSourceDirectory sourceFolder : this.sourceFolders) {
+            try {
+                sourceFolderEntries.add(createSourceFolderEntry(sourceFolder, rawClasspath));
+            } catch (Exception e) {
+                CorePlugin.logger().warn("Failed to create source entry", e);
```
Comment (etiennestuder) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/SourceFolderUpdater.java]: Why do we catch this exception here and then continue?

File: org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/SourceFolderUpdater.java [hunk 2]
```diff
@@ -76,43 +80,70 @@ private void updateClasspath(IProgressMonitor monitor) throws JavaModelException
...
+        // duplicate source folders
+        return ImmutableSet.copyOf(sourceFolderEntries.build()).asList();
+    }
 
-                // find the source folder among the sources currently configured on the project
-                Optional<IClasspathEntry> currentClasspathEntry = FluentIterable.from(rawClasspath).firstMatch(new Predicate<IClasspathEntry>() {
+    private IClasspathEntry createSourceFolderEntry(OmniEclipseSourceDirectory directory, List<IClasspathEntry> rawClasspath) throws CoreException {
+        // TODO (donat) the creation of the linked folder has to be created before the code
+        // below is executed. should we do something about it?
+        Optional<IFolder> linkedFolder = getLinkedFolderIfExists(directory.getDirectory());
```
Comment (etiennestuder) [inline org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/SourceFolderUpdater.java]: Shouldn't the LinkedResourcesUpdater take care of this?

**Pull request merged**